crates/biome_css_parser/src/syntax/at_rule/tailwind.rs

```diff
@@ -0,0 +1,240 @@
+        // Functional utility: tab-*
+        let m = p.start();
+
+        parse_regular_identifier(p).ok();
+        p.expect(T![-]);
+        p.expect(T![*]);
+
+        Present(m.complete(p, TW_FUNCTIONAL_UTILITY_NAME))
+    } else {
+        // Simple utility: center-flex
+        parse_regular_identifier(p)
+    }
+}
+
+// @variant dark { background: black; }
+pub(crate) fn parse_variant_at_rule(p: &mut CssParser) -> ParsedSyntax {
+    if !p.at(T![variant]) {
+        return Absent;
+    }
+
+    let m = p.start();
+    p.bump(T![variant]);
+
+    if !is_at_identifier(p) {
+        p.error(expected_identifier(p, p.cur_range()));
+        return Present(m.complete(p, CSS_BOGUS_AT_RULE));
+    }
+
+    parse_regular_identifier(p).ok();
+
+    parse_declaration_or_rule_list_block(p);
+
+    Present(m.complete(p, TW_VARIANT_AT_RULE))
+}
+
+// @custom-variant theme-midnight (&:where([data-theme="midnight"] *));
+pub(crate) fn parse_custom_variant_at_rule(p: &mut CssParser) -> ParsedSyntax {
+    if !p.at(T![custom_variant]) {
+        return Absent;
+    }
+
+    let m = p.start();
+    p.bump(T![custom_variant]);
+
+    if !is_at_identifier(p) {
+        p.error(expected_identifier(p, p.cur_range()));
+        return Present(m.complete(p, CSS_BOGUS_AT_RULE));
+    }
+
+    parse_regular_identifier(p).ok();
+
+    if p.at(T!['(']) {
+        // shorthand syntax
+        // @custom-variant theme-midnight (&:where([data-theme="midnight"] *));
+        parse_custom_variant_shorthand(p).ok();
+    } else {
+        // longhand syntax
+        // @custom-variant theme-midnight {
+        //   &:where([data-theme="midnight"] *) {
+        //     @slot;
+        //   }
+        // }
+        parse_rule_block(p);
+    }
+
+    Present(m.complete(p, TW_CUSTOM_VARIANT_AT_RULE))
+}
+
+fn parse_custom_variant_shorthand(p: &mut CssParser) -> ParsedSyntax {
+    if !p.at(T!['(']) {
+        return Absent;
+    }
+
+    let m = p.start();
+
+    p.bump(T!['(']);
+    parse_selector(p).ok();
+    p.expect(T![')']);
+    p.expect(T![;]);
+
+    Present(m.complete(p, TW_CUSTOM_VARIANT_SHORTHAND))
+}
+
+// @apply text-lg font-bold;
+pub(crate) fn parse_apply_at_rule(p: &mut CssParser) -> ParsedSyntax {
+    if !p.at(T![apply]) {
+        return Absent;
+    }
+
+    let m = p.start();
+    p.bump_with_context(T![apply], CssLexContext::TailwindUtility);
+    ApplyClassList.parse_list(p);
+    p.expect(T![;]);
+
+    Present(m.complete(p, TW_APPLY_AT_RULE))
+}
+
+struct ApplyClassList;
+
+impl ParseNodeList for ApplyClassList {
+    type Kind = CssSyntaxKind;
+    type Parser<'source> = CssParser<'source>;
+    const LIST_KIND: Self::Kind = TW_APPLY_CLASS_LIST;
+
+    fn parse_element(&mut self, p: &mut Self::Parser<'_>) -> ParsedSyntax {
+        parse_identifier(p, CssLexContext::TailwindUtility)
+    }
+
+    fn is_at_list_end(&self, p: &mut Self::Parser<'_>) -> bool {
+        p.at(T![;]) || p.at(EOF)
```

Comment:
```suggestion
        p.at(T![;])
```

I remember that  `ParseNodeList` should EOF 🤔